pkg/ovsdb/transact.go

```diff
@@ -0,0 +1,1094 @@
+	}
+	fn := c.Function
+	expected, ok := c.Value.(libovsdb.OvsSet)
+	if !ok {
+		return false, errors.New(E_CONSTRAINT_VIOLATION)
+	}
+
+	if (fn == FN_EQ || fn == FN_IN) && isEqualSet(actual, expected) {
+		return true, nil
+	}
+	if (fn == FN_NE || fn == FN_EX) && !isEqualSet(actual, expected) {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (c *Condition) CompareMap(row *map[string]interface{}) (bool, error) {
+	actual, ok := (*row)[c.Column].(libovsdb.OvsMap)
+	if !ok {
+		return false, errors.New(E_CONSTRAINT_VIOLATION)
+	}
+	fn := c.Function
+	expected, ok := c.Value.(libovsdb.OvsMap)
+	if !ok {
+		return false, errors.New(E_CONSTRAINT_VIOLATION)
+	}
+
+	if (fn == FN_EQ || fn == FN_IN) && isEqualMap(actual, expected) {
+		return true, nil
+	}
+	if (fn == FN_NE || fn == FN_EX) && !isEqualMap(actual, expected) {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (c *Condition) Compare(row *map[string]interface{}) (bool, error) {
+	switch (*row)[c.Column].(type) {
+	case int:
+		return c.CompareInteger(row)
+	case float64:
+		return c.CompareReal(row)
+	case bool:
+		return c.CompareBoolean(row)
+	case string:
+		return c.CompareString(row)
+	case libovsdb.OvsSet:
+		return c.CompareSet(row)
+	case libovsdb.OvsMap:
+		return c.CompareMap(row)
+	case libovsdb.UUID:
+		return c.CompareUUID(row)
+	default:
+		return false, errors.New(E_CONSTRAINT_VIOLATION)
+	}
+}
+
+func isCondTypeUUID(cond1 interface{}) (string, error) {
+	cond2, ok := cond1.([]interface{})
+	if !ok {
+		return "", errors.New(E_ERROR)
+	}
+	condition, err := NewCondition(cond2)
+	if err != nil {
+		return "", err
+	}
+	if condition.Column != COL_UUID {
+		return "", nil
+	}
+	if condition.Function != FN_EQ {
+		return "", nil
+	}
+	ovsUUID, ok := condition.Value.(libovsdb.UUID)
+	if !ok {
+		return "", errors.New(E_ERROR)
+	}
+	_, err = guuid.Parse(ovsUUID.GoUUID)
+	if err != nil {
+		return "", err
+	}
+	return ovsUUID.GoUUID, err
+}
+
+func doesWhereContainCondTypeUUID(where []interface{}) (string, error) {
+	for _, c := range where {
+		cond, ok := c.([]interface{})
+		if !ok {
+			return "", errors.New(E_ERROR)
+		}
+		uuid, err := isCondTypeUUID(cond)
+		if err != nil {
+			return "", err
+		}
+		if uuid != "" {
+			return uuid, nil
+		}
+	}
+	return "", nil
+
+}
+
+func isRowSelectedByWhere(row *map[string]interface{}, where []interface{}) (bool, error) {
+	for _, c := range where {
+		cond, ok := c.([]interface{})
+		if !ok {
+			return false, errors.New(E_ERROR)
+		}
+		ok, err := isRowSelectedByCond(row, cond)
+		if err != nil {
+			return false, err
+		}
+		if !ok {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func isRowSelectedByCond(row *map[string]interface{}, cond []interface{}) (bool, error) {
+	condition, err := NewCondition(cond)
+	if err != nil {
+		return false, err
+	}
+	return condition.Compare(row)
+}
+
+func reduceRowByColumns(row *map[string]interface{}, columns []string) *map[string]interface{} {
```

Comment:
I have similar functionality in Monitors

Reply:
ok